contrib/database/sql/conn.go

```diff
@@ -135,7 +138,12 @@ func (tc *TracedConn) ExecContext(ctx context.Context, query string, args []driv
 		cquery, spanID := tc.injectComments(ctx, query, tc.cfg.dbmPropagationMode)
 		ctx, end := startTraceTask(ctx, QueryTypeExec)
 		defer end()
-		r, err := execContext.ExecContext(ctx, cquery, args)
+		if appsec.Enabled() {
+			err = sqlsec.ProtectSQLOperation(ctx, query, tc.driverName)
+		}
+		if !events.IsSecurityError(err) {
+			r, err = execContext.ExecContext(ctx, cquery, args)
+		}
```

Comment:
```suggestion
		if shouldExecuteQuery(ctx, query, tc.driverName)
			r, err = execContext.ExecContext(ctx, cquery, args)
		}
```

I think this could be simplified by extracting this logic to its own function:

```go
func shouldExecuteQuery(ctx context.Context, query, driverName string) bool {
	if !appsec.Enabled() {
		return true
	}
	err := sqlsec.ProtectSQLOperation(ctx, query, driverName)
	return !events.IsSecurityError(err)
}
```

Also, since when the err is not a security error, you are replacing it with the error from `r, err = execContext.ExecContext(ctx, cquery, args)`, with this change you could get rid of the `!events.IsSecurityError(err)` check from line 345.

Please let me know if this makes sense!

Reply:
We still need the error to be passed to `tryTrace` so that in case of security error it doesn't get tagged.
I put the appsec.Enabled() + security monitoring call in its own function, but made it return the error so that we keep track of it for that purpose. See be265f750c2cb851cb

As for testing, this happens in the appsec pkg (see `waf_tests.go` in the PR) since the added functionalities don't impact tracer execution when running with APM only and are related to ASM (query blocking, span tagging, stacktrace generation, etc...)

Reply:
> We still need the error to be passed to tryTrace so that in case of security error it doesn't get tagged.
With my suggestion, err would be `nil` in that case, so it would not be tagged.